x/ccv/provider/keeper/relay.go

```diff
@@ -190,16 +162,45 @@ func (k Keeper) SendPendingVSCPackets(ctx sdk.Context, chainID, channelID string
 			channelID,          // source channel id
 			ccv.ProviderPortID, // source port id
 			data.GetBytes(),
-			k.GetParams(ctx).CcvTimeoutPeriod,
+			k.GetCCVTimeoutPeriod(ctx),
 		)
+
 		if err != nil {
+			if clienttypes.ErrClientNotActive.Is(err) {
+				// IBC client is expired!
+				// leave the packet data stored to be sent once the client is upgraded
+				// the client cannot expire during iteration (in the middle of a block)
+				return
+			}
 			panic(fmt.Errorf("packet could not be sent over IBC: %w", err))
 		}
 		// set the VSC send timestamp for this packet;
 		// note that the VSC send timestamp are set when the packets
 		// are actually sent over IBC
 		k.SetVscSendTimestamp(ctx, chainID, data.ValsetUpdateId, ctx.BlockTime())
 	}
+	k.DeletePendingPackets(ctx, chainID)
+}
+
+// QueueVSCPackets queues latest validator updates for every registered consumer chain
+func (k Keeper) QueueVSCPackets(ctx sdk.Context) {
+	valUpdateID := k.GetValidatorSetUpdateId(ctx) // curent valset update ID
+	// check whether there are changes in the validator set;
+	valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)
+
+	k.IterateConsumerChains(ctx, func(ctx sdk.Context, chainID, clientID string) (stop bool) {
+		// note that this also entails unbonding operations
+		// w/o changes in the voting power of the validators in the validator set
+		unbondingOps, _ := k.GetUnbondingOpsFromIndex(ctx, chainID, valUpdateID)
+		if len(valUpdates) != 0 || len(unbondingOps) != 0 {
+			// construct validator set change packet data
+			packet := ccv.NewValidatorSetChangePacketData(valUpdates, valUpdateID, k.ConsumeSlashAcks(ctx, chainID))
+			k.AppendPendingPackets(ctx, chainID, packet)
+		}
+		return true // do not stop the iteration
+	})
+
+	k.IncrementValidatorSetUpdateId(ctx)
```

Comment:
Take this line out from `QueueVSCPackets`. Makes more sense to be directly in `EndBlockVSU`.

Reply:
It's convienient to keep it there since it makes unittesting the function easier.

The unittest checks that the updateId was incremented. It's a lot more involved to test the `EndBlockVSU` to see if the updateId got incremented after queueing.

It is also makes the logic easier to follow (just IMHO). You can see immediately that the counter is incremented inside `QueueVSCPackets` and don't have to check other functions called in `EndBlock`.

I'll refactor if you feel strongly about it.

Reply:
Okay. Let's leave it then.